// Copyright (c) 2022 Samsung Research Russia
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "nav2_collision_monitor/collision_monitor_node.hpp"

#include "safety_monitor_msgs/msg/field_states.hpp"

#include <exception>
#include <utility>
#include <functional>

#include "nav2_util/node_utils.hpp"

namespace nav2_collision_monitor
{

CollisionMonitor::CollisionMonitor(const rclcpp::NodeOptions & options)
: nav2_util::LifecycleNode("collision_monitor", "", false, options),
  process_active_(false)
{
}

CollisionMonitor::~CollisionMonitor()
{
  for (auto source : sources_) {
    source->polygons.clear();
  }
  sources_.clear();
}

nav2_util::CallbackReturn
CollisionMonitor::on_configure(const rclcpp_lifecycle::State & /*state*/)
{
  RCLCPP_INFO(get_logger(), "Configuring");

  // Obtaining ROS parameters
  if (!getParameters()) {
    return nav2_util::CallbackReturn::FAILURE;
  }

  timer_ = rclcpp::create_timer(
    this, this->get_clock(), std::chrono::milliseconds(1),
    std::bind(&CollisionMonitor::process, this));
  return nav2_util::CallbackReturn::SUCCESS;
}

nav2_util::CallbackReturn
CollisionMonitor::on_activate(const rclcpp_lifecycle::State & /*state*/)
{
  RCLCPP_INFO(get_logger(), "Activating");

  // Activating polygons
  for (auto source : sources_) {
    for (auto polygon : source->polygons) {
      polygon->activate();
    }
  }

  // Since polygons are being published when a scan appears,
  // we need to publish polygons first time to display them at startup
  publishPolygons();

  // Activating main worker
  process_active_ = true;

  // Creating bond connection
  createBond();

  return nav2_util::CallbackReturn::SUCCESS;
}

nav2_util::CallbackReturn
CollisionMonitor::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
{
  RCLCPP_INFO(get_logger(), "Deactivating");

  // Deactivating main worker
  process_active_ = false;

  // Deactivating polygons
  for (auto source : sources_) {
    for (auto polygon : source->polygons) {
      polygon->deactivate();
    }
  }

  // Destroying bond connection
  destroyBond();

  return nav2_util::CallbackReturn::SUCCESS;
}

nav2_util::CallbackReturn
CollisionMonitor::on_cleanup(const rclcpp_lifecycle::State & /*state*/)
{
  RCLCPP_INFO(get_logger(), "Cleaning up");

  for (auto source : sources_) {
    source->polygons.clear();
  }
  sources_.clear();

  return nav2_util::CallbackReturn::SUCCESS;
}

nav2_util::CallbackReturn
CollisionMonitor::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
{
  RCLCPP_INFO(get_logger(), "Shutting down");

  return nav2_util::CallbackReturn::SUCCESS;
}

bool CollisionMonitor::getParameters()
{
  rclcpp::Duration source_timeout(2.0, 0.0);

  auto node = shared_from_this();

  nav2_util::declare_parameter_if_not_declared(
    node, "source_timeout", rclcpp::ParameterValue(2.0));
  source_timeout =
    rclcpp::Duration::from_seconds(get_parameter("source_timeout").as_double());

  if (!configureSources(source_timeout)) {
    return false;
  }

  return true;
}

bool CollisionMonitor::configureSourcePolygons(
  std::shared_ptr<Source> source,
  std::string source_name)
{
  auto node = shared_from_this();

  nav2_util::declare_parameter_if_not_declared(
    node, source_name + ".polygons", rclcpp::ParameterValue(std::vector<std::string>()));
  std::vector<std::string> polygon_names =
    get_parameter(source_name + ".polygons").as_string_array();
  for (std::string polygon_name : polygon_names) {
    polygon_name = source_name + "." + polygon_name;
    nav2_util::declare_parameter_if_not_declared(
      node, polygon_name + ".type", rclcpp::PARAMETER_STRING);
    const std::string polygon_type = get_parameter(polygon_name + ".type").as_string();

    if (polygon_type == "polygon") {
      source->polygons.push_back(
        std::make_shared<Polygon>(
          node, polygon_name));
    } else if (polygon_type == "circle") {
      source->polygons.push_back(
        std::make_shared<Circle>(
          node, polygon_name));
    } else {     // Error if something else
      RCLCPP_ERROR(
        get_logger(),
        "[%s]: Unknown polygon type: %s",
        polygon_name.c_str(), polygon_type.c_str());
      return false;
    }

    // Configure last added polygon
    if (!source->polygons.back()->configure()) {
      return false;
    }
  }
  return true;
}

bool CollisionMonitor::configureSources(const rclcpp::Duration & source_timeout)
{
  try {
    auto node = shared_from_this();

    // Leave it to be not initialized: to intentionally cause an error if it will not set
    nav2_util::declare_parameter_if_not_declared(
      node, "observation_sources", rclcpp::PARAMETER_STRING_ARRAY);
    std::vector<std::string> source_names = get_parameter("observation_sources").as_string_array();
    for (std::string source_name : source_names) {
      nav2_util::declare_parameter_if_not_declared(
        node, source_name + ".type",
        rclcpp::ParameterValue("scan"));  // Laser scanner by default
      const std::string source_type = get_parameter(source_name + ".type").as_string();

      if (source_type == "scan") {
        std::shared_ptr<Scan> s = std::make_shared<Scan>(
          node, source_name, source_timeout);
        s->configure();
        configureSourcePolygons(s, source_name);
        sources_.push_back(s);
      } else if (source_type == "pointcloud") {
        std::shared_ptr<PointCloud> p = std::make_shared<PointCloud>(
          node, source_name, source_timeout);
        p->configure();
        configureSourcePolygons(p, source_name);
        sources_.push_back(p);
      } else {  // Error if something else
        RCLCPP_ERROR(
          get_logger(),
          "[%s]: Unknown source type: %s",
          source_name.c_str(), source_type.c_str());
        return false;
      }
    }
  } catch (const std::exception & ex) {
    RCLCPP_ERROR(get_logger(), "Error while getting parameters: %s", ex.what());
    return false;
  }

  return true;
}

void CollisionMonitor::process()
{
  // Current timestamp for all inner routines prolongation
  rclcpp::Time curr_time = this->now();

  // Do nothing if main worker in non-active state
  if (!process_active_) {
    return;
  }

  // Points array collected from different data sources in a robot base frame
  std::vector<Point> collision_points;

  // Fill collision_points array from different data sources
  for (std::shared_ptr<Source> source : sources_) {
    auto msg = safety_monitor_msgs::msg::FieldStates();
    source->getData(curr_time, collision_points);

    for (auto polygon : source->polygons) {
      msg.names.push_back(polygon->getName());
      msg.triggered.push_back(polygon->getPointsInside(collision_points) > polygon->getMaxPoints());
    }
    source->pub_.publish(msg);
  }
  publishPolygons();
}

void CollisionMonitor::publishPolygons() const
{
  for (auto source : sources_) {
    for (auto polygon : source->polygons) {
      polygon->publish();
    }
  }
}

}  // namespace nav2_collision_monitor

#include "rclcpp_components/register_node_macro.hpp"

// Register the component with class_loader.
// This acts as a sort of entry point, allowing the component to be discoverable when its library
// is being loaded into a running process.
RCLCPP_COMPONENTS_REGISTER_NODE(nav2_collision_monitor::CollisionMonitor)
